[Review] fix(eventloop): Fix deadlock with event loop mutex

keba-estr opened the pull request:
As specific connection is not known yet, lock event loop mutexes of
all connections.
When adding and removing a connection from the server list, lock/unlock
the mutex to match the locking counter.

Resolves https://github.com/open62541/open62541/issues/7074

jpfr commented:
Thanks for this.
PR #7085 is a bit more lightweight.
Does that work for you also?

keba-estr commented:
> Thanks for this. PR #7085 is a bit more lightweight. Does that work for you also?

No, PR #7085 is not enough, it does not work for me.
The problem is that the dc callback is still called with the locked mutex, there are no changes for the called functions in UA_PubSubConnection_disconnect or processDelayed with that PR.

Also, if a separate eventloop is used for PubSub, what is the cause for also locking/unlocking the server eventloop mutex with the functions lockServer/unlockServer in the PubSub files?

keba-estr commented:
See discussion in issue #7074

keba-estr committed:
fix(eventloop): lock PubSub eventloop mutex

As specific connection is not known yet, lock event loop mutexes of
all connections.
When adding and removing a connection from the server list, lock/unlock
the mutex to match the locking counter.

keba-estr commented:
@jpfr: What is the cause of the failing builds?

jpfr commented:
Two quick comments:

- You can rename ua_pubsub.h to ua_pubsub_internal.h. It is used exactly like that already.
- There is a lot of repetition in the added code. You could pull it into a commonly used function.
For example ``lockPubSubServerWithConnection``. Where the EventLoop to be used is selected inside that function.

keba-estr commented:
> Two quick comments:
> 
> * You can rename ua_pubsub.h to ua_pubsub_internal.h. It is used exactly like that already.
> * There is a lot of repetition in the added code. You could pull it into a commonly used function.
>   For example `lockPubSubServerWithConnection`. Where the EventLoop to be used is selected inside that function.

I would prefer to add the functions to ua_pubsub.h then without renaming any files. This minimizes the needed changes.

keba-estr commented:
The order of the mutexes is still not used consistently.
In lockPubSubServer the event loop mutex is locked and unlocked before the server mutex.
Yet the function callstack UA_Server_run -> UA_Server_run_shutdown -> UA_Server_run_shutdown -> UA_PubSubManager_shutdown -> UA_PubSubConnection_setPubSubState -> UA_PubSubConnection_disconnect ->UDP_shutdownConnection
first locks the server mutex in UA_Server_run_shutdown and then the event loop mutex in UDP_shutdownConnection, which sill causes a deadlock.

keba-estr committed:
fix(pubsub): assure locking order

Always lock event loop lock before server lock

jpfr commented:
TODO: Review existing PR after discussion with @keba-estr and @keba-ebr.

```
	1.	PubSub EventLoop Lock
	2.	Server EventLoop Lock
	3.	Server Lock (ServiceMutex)

1. Idea: Fixed order for taking all locks

2. Idea: PubSub EventLoop gets released before calling into the server
```

keba-estr commented:
According to Mr. Julius Pfrommer usage of a separate PubSub event loop is a special case which is only used by a small number of companies, there will be no effort taken for this issue.
This problem does not appear on the 1.5 brach as the specific event loop has been removed and the user is responsible for handling the customstatemachine.
